File: python/Linked List/obj.py
class Node:
    def __init__(self, data):
        self.data = data      
        self.next = None      

class LinkedList:
    def __init__(self):
        self.head = None      

    def append(self, data):
        new_node = Node(data)

        if self.head is None:
            self.head = new_node
            return

        last_node = self.head
        while last_node.next:
            last_node = last_node.next

        last_node.next = new_node

    def prepend(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def delete_node(self, key):
        current_node = self.head

     
        if current_node and current_node.data == key:
            self.head = current_node.next
            current_node = None
            return

     
        prev = None
        while current_node and current_node.data != key:
            prev = current_node
            current_node = current_node.next

        if current_node is None:
            print(f"Node with data {key} not found.")
            return

        prev.next = current_node.next
        current_node = None


    def display(self):
        current_node = self.head
        while current_node:
            print(current_node.data, end=" -> ")
            current_node = current_node.next
        print("end")
    def delete_first(self):
        
        if self.head is None:
            print("empty")
            return

       
        self.head = self.head.next
    def delete_last(self):
        
        if self.head is None:
            print("List is empty")
            return

        
        if self.head.next is None:
            self.head = None
            return

        
        second_last = self.head
        while second_last.next.next:
            second_last = second_last.next

        
        second_last.next = None
        def delete_at_index(self, index):

            if self.head is None:
                print("List is empty")
                return

            if index == 0:
                self.head = self.head.next
                return

            current = self.head
            position = 0


            while current is not None and position < index - 1:
                current = current.next
                position += 1

          
            if current is None or current.next is None:
                print("Index out of bounds")
                return


            current.next = current.next.next

File: c/Linked List/obj.c
#include <stdio.h>
#include <stdlib.h>
#include "obj.h"
Node* create_node(int data) {
    Node* new_node = (Node*)malloc(sizeof(Node));
    if (new_node == NULL) {
        fprintf(stderr, "Error: Memory allocation failed.\n");
        return NULL;
    }
    new_node->data = data;
    new_node->next = NULL;
    return new_node;
}
void insert_at_head(Node** head, int data) {
    Node* new_node = create_node(data);
    if (new_node == NULL) return;
    new_node->next = *head; 
    *head = new_node;       
}
void insert_at_tail(Node** head, int data) {
    Node* new_node = create_node(data);
    if (new_node == NULL) return;
    if (*head == NULL) {
        *head = new_node;
        return;
    }
    Node* temp = *head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = new_node;
}
void delete_node(Node** head, int key) {
    if (*head == NULL) return;
    Node* temp = *head;
    Node* prev = NULL;
    if (temp != NULL && temp->data == key) {
        *head = temp->next; 
        free(temp);         
        return;
    }
    while (temp != NULL && temp->data != key) {
        prev = temp;
        temp = temp->next;
    }
    if (temp == NULL) return;
    prev->next = temp->next;
    free(temp);
}
void print_list(Node* head) {
    Node* temp = head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}
void free_list(Node* head) {
    Node* temp = head;
    while (temp != NULL) {
        Node* next_node = temp->next; 
        free(temp);                   
        temp = next_node;             
    }
}
